### PR TITLE
Fix memory visualization, with a detour through coord

### PR DESCRIPTION
**Pull request by benesch**
The upshot here is that the memory visualization is fixed. This required a long detour through `coord` to avoid using `NoSessionExecute`, which required sharing the ID allocation/secret key infrastructure between the HTTP and pgwire servers, which just turned into a lot of code movement. Things should be pretty nicely split up by commit though.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5746)

### CODE REVIEW

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -1006,13 +1043,20 @@ impl Coordinator {
...
     /// the named `conn_id`.
-    async fn handle_cancel(&mut self, conn_id: u32) {
-        // Tell dataflow to cancel any pending peeks.
-        self.broadcast(SequencedCommand::CancelPeek { conn_id });
+    async fn handle_cancel(&mut self, conn_id: u32, secret_key: u32) {
+        if let Some(conn_meta) = self.active_conns.get(&conn_id) {
+            // If the secret key specified by the client doesn't match the
+            // actual secret key for the target connection, we treat this as a
+            // rogue cancellation request and ignore it.
+            if conn_meta.secret_key == secret_key {
```
Comment (mjibson) [inline src/coord/src/coord.rs:1012]: `!=`?

File: src/materialized/src/http.rs [hunk 2]
```diff
@@ -140,30 +141,44 @@ impl Server {
...
+            let coord_client = self.coord_client.clone();
+            let start_time = self.start_time;
+            async move {
+                let user = match user {
+                    Ok(user) => user,
+                    Err(e) => return Ok(util::error_response(StatusCode::UNAUTHORIZED, e)),
+                };
+
+                let coord_client = coord_client.new_conn()?;
+                let session = Session::new(coord_client.conn_id(), user);
```
Comment (mjibson) [inline src/materialized/src/http.rs:154]: If we are creating sessions in more than one place, I think doing the "session's drop impl checks that terminate was called" thing is almost certainly the best thing to do. Especially if we may also support TAIL and other stuff over http in the future.

File: src/materialized/src/http.rs [hunk 3]
```diff
@@ -140,30 +141,44 @@ impl Server {
...
+                    }
+                    (&Method::GET, "/prof") => prof::handle_prof(req, coord_client).await,
+                    (&Method::GET, "/memory") => memory::handle_memory(req, coord_client).await,
+                    (&Method::POST, "/prof") => prof::handle_prof(req, coord_client).await,
+                    (&Method::POST, "/sql") => sql::handle_sql(req, coord_client).await,
+                    (&Method::GET, "/internal/catalog") => {
+                        catalog::handle_internal_catalog(req, coord_client).await
+                    }
+                    _ => root::handle_static(req, coord_client).await,
                 }
```
Comment (mjibson) [inline src/materialized/src/http.rs:181]: Doesn't the session need to be terminated here?

Comment (benesch) [reply src/materialized/src/http.rs:181]: 🙈

File: src/coord/src/client.rs [hunk 4]
```diff
@@ -120,15 +128,14 @@ impl Client {
 }
 
 /// A [`Client`] that is bound to a session.
+///
+/// You must call [`SessionClient::terminate`] rather than dropping a session
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/coord/src/client.rs:160]: Definitely not overdoing it considering how immediately I got this wrong :D. Super easy to do; `self.session` is already this sentinel.

File: src/coord/src/coord.rs (hunk 1 above)
Comment (benesch) [reply src/coord/src/coord.rs:1012]: Oy, thanks! Luckily there was plenty of test coverage for this.

**Comment by benesch** [comment]
Thanks for spotting those bugs. Fixed as you suggestion. Hoping for happier test outcomes this time around.

**Review by mjibson — APPROVED** [review]

**Pull request merged**